### PR TITLE
Add grid stride tunings of reduction kernels

### PR DESCRIPTION
**Pull request by MrBurmark**
# Add grid stride tunings of reduction kernels

Add a occupancy calculator grid stride (occgs_<block_size>) tuning for Cuda and Hip variants of reduction kernels.

This generally improves performance at problem sizes greater than the maximum occupancy of the device. This happens because the amount of work to finalize the reduction is proportional to the number of blocks.

The occupancy calculator grid stride tuning works by launching fewer total threads than iterates and uses a grid stride loop to assign multiple iterates to each thread. The maximum number of blocks is determined by the occupancy calculator to maximize occupancy.

- This PR is a feature
- It does the following:
  - Adds better GPU reduction tunings at the request of myself

### CODE REVIEW

File: src/algorithm/REDUCE_SUM-Cuda.cpp [hunk 1]
```diff
@@ -190,11 +192,79 @@ void REDUCE_SUM::runCudaVariantBlock(VariantID vid)
...
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      cudaErrchk( cudaMemcpyAsync( dsum, &m_sum_init, sizeof(Real_type),
+                                   cudaMemcpyHostToDevice, res.get_stream() ) );
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      reduce_sum<block_size><<<grid_size, block_size,
+                  shmem, res.get_stream()>>>( x,
+                                                   dsum, m_sum_init,
```
Comment (rhornung67) [inline src/algorithm/REDUCE_SUM-Cuda.cpp:225]: vertical alignment of args would look better here.

**Review by rhornung67 — APPROVED** [review]

File: src/algorithm/REDUCE_SUM-Cuda.cpp (hunk 1 above)
Comment (artv3) [reply src/algorithm/REDUCE_SUM-Cuda.cpp:225]: agreed

**Review by artv3 — APPROVED** [review]

File: src/algorithm/REDUCE_SUM-Cuda.cpp [hunk 2]
```diff
@@ -190,11 +192,79 @@ void REDUCE_SUM::runCudaVariantBlock(VariantID vid)
...
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      cudaErrchk( cudaMemcpyAsync( dsum, &m_sum_init, sizeof(Real_type),
+                                   cudaMemcpyHostToDevice, res.get_stream() ) );
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      reduce_sum<block_size><<<grid_size, block_size,
+                  shmem, res.get_stream()>>>( x,
```
Comment (MrBurmark) [inline src/algorithm/REDUCE_SUM-Cuda.cpp:224]:
```suggestion
      reduce_sum<block_size><<<grid_size, block_size,
                               shmem, res.get_stream()>>>( x,
```

File: src/basic/PI_REDUCE-Cuda.cpp [hunk 3]
```diff
@@ -120,7 +122,132 @@ void PI_REDUCE::runCudaVariantImpl(VariantID vid)
...
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      cudaErrchk( cudaMemcpyAsync( dpi, &m_pi_init, sizeof(Real_type),
+                                   cudaMemcpyHostToDevice, res.get_stream() ) );
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      pi_reduce<block_size><<<grid_size, block_size,
+                  shmem, res.get_stream()>>>( dx,
```
Comment (MrBurmark) [inline src/basic/PI_REDUCE-Cuda.cpp:154]:
```suggestion
      pi_reduce<block_size><<<grid_size, block_size,
                              shmem, res.get_stream()>>>( dx,
```

File: src/basic/REDUCE3_INT-Cuda.cpp [hunk 4]
```diff
@@ -151,7 +153,148 @@ void REDUCE3_INT::runCudaVariantImpl(VariantID vid)
...
+      vmem_init[0] = m_vsum_init;
+      vmem_init[1] = m_vmin_init;
+      vmem_init[2] = m_vmax_init;
+      cudaErrchk( cudaMemcpyAsync( vmem, vmem_init, 3*sizeof(Int_type),
+                                   cudaMemcpyHostToDevice, res.get_stream() ) );
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      reduce3int<block_size><<<grid_size, block_size,
+                   shmem, res.get_stream()>>>(vec,
```
Comment (MrBurmark) [inline src/basic/REDUCE3_INT-Cuda.cpp:191]:
```suggestion
      reduce3int<block_size><<<grid_size, block_size,
                               shmem, res.get_stream()>>>(vec,
```

File: src/basic/REDUCE3_INT-Hip.cpp [hunk 5]
```diff
@@ -151,7 +153,149 @@ void REDUCE3_INT::runHipVariantImpl(VariantID vid)
...
+
+      vmem_init[0] = m_vsum_init;
+      vmem_init[1] = m_vmin_init;
+      vmem_init[2] = m_vmax_init;
+      hipErrchk( hipMemcpyAsync( vmem, vmem_init, 3*sizeof(Int_type),
+                                 hipMemcpyHostToDevice, res.get_stream() ) );
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      hipLaunchKernelGGL((reduce3int<block_size>), dim3(grid_size), dim3(block_size), shmem, res.get_stream(),
```
Comment (MrBurmark) [inline src/basic/REDUCE3_INT-Hip.cpp:190]:
```suggestion
      hipLaunchKernelGGL((reduce3int<block_size>), dim3(grid_size), dim3(block_size),
                                                   shmem, res.get_stream(),
```

File: src/basic/REDUCE_STRUCT-Hip.cpp [hunk 6]
```diff
@@ -180,8 +182,155 @@ void REDUCE_STRUCT::runHipVariantImpl(VariantID vid)
...
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      hipErrchk(hipMemsetAsync(mem, 0.0, 6*sizeof(Real_type), res.get_stream()));
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      hipLaunchKernelGGL((reduce_struct<block_size>),
+                         dim3(grid_size), dim3(block_size),
+                         shmem, res.get_stream(),
+                   points.x, points.y,
```
Comment (MrBurmark) [inline src/basic/REDUCE_STRUCT-Hip.cpp:215]:
```suggestion
                         shmem, res.get_stream(),
                         points.x, points.y,
```

File: src/basic/TRAP_INT-Cuda.cpp [hunk 7]
```diff
@@ -143,7 +145,135 @@ void TRAP_INT::runCudaVariantImpl(VariantID vid)
...
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      cudaErrchk( cudaMemcpyAsync( sumx, &m_sumx_init, sizeof(Real_type),
+                                   cudaMemcpyHostToDevice, res.get_stream() ) );
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      trapint<block_size><<<grid_size, block_size,
+                shmem, res.get_stream()>>>(x0, xp,
```
Comment (MrBurmark) [inline src/basic/TRAP_INT-Cuda.cpp:177]:
```suggestion
      trapint<block_size><<<grid_size, block_size,
                            shmem, res.get_stream()>>>(x0, xp,
```

File: src/basic/TRAP_INT-Hip.cpp [hunk 8]
```diff
@@ -142,7 +144,135 @@ void TRAP_INT::runHipVariantImpl(VariantID vid)
...
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      hipErrchk( hipMemcpyAsync( sumx, &m_sumx_init, sizeof(Real_type),
+                                 hipMemcpyHostToDevice, res.get_stream() ) );
+
+      const size_t normal_grid_size = RAJA_DIVIDE_CEILING_INT(iend, block_size);
+      const size_t grid_size = std::min(normal_grid_size, max_grid_size);
+      hipLaunchKernelGGL((trapint<block_size>), dim3(grid_size), dim3(block_size), shmem, res.get_stream(), x0, xp,
```
Comment (MrBurmark) [inline src/basic/TRAP_INT-Hip.cpp:175]:
```suggestion
      hipLaunchKernelGGL((trapint<block_size>), dim3(grid_size), dim3(block_size),
                                                shmem, res.get_stream(), x0, xp,
```

File: src/algorithm/REDUCE_SUM-Cuda.cpp (hunk 1 above)
Comment (MrBurmark) [reply src/algorithm/REDUCE_SUM-Cuda.cpp:225]: improved

Comment (MrBurmark) [reply src/algorithm/REDUCE_SUM-Cuda.cpp:225]: We should go through everything and improve the spacing of kernel launch args. We often have the first kernel argument at the end of the first line with the launch arguments which is confusing.

Comment (MrBurmark) [reply src/algorithm/REDUCE_SUM-Cuda.cpp:225]: https://github.com/LLNL/RAJAPerf/issues/371

**Pull request merged**